lua/telescope/pickers/highlights.lua

```diff
@@ -56,65 +61,106 @@ end
 
 function Highlighter:hi_sorter(row, prompt, display)
   local picker = self.picker
+  local sorter = picker.sorter
   if not picker.sorter or not picker.sorter.highlighter then
     return
   end
 
   local results_bufnr = assert(self.picker.results_bufnr, "Must have a results bufnr")
-  picker:highlight_one_row(results_bufnr, prompt, display, row)
+  if not a.nvim_buf_is_valid(results_bufnr) then
+    return
+  end
+
+  local sorter_highlights = sorter:highlighter(prompt, display)
+
+  if sorter_highlights then
+    for _, hl in ipairs(sorter_highlights) do
+      local highlight, start, finish
+      if type(hl) == "table" then
+        highlight = hl.highlight or "TelescopeMatching"
+        start = hl.start
+        finish = hl.finish or hl.start
+      elseif type(hl) == "number" then
+        highlight = "TelescopeMatching"
+        start = hl
+        finish = hl
+      else
+        error "Invalid highlight"
+      end
+
+      a.nvim_buf_set_extmark(results_bufnr, ns_telescope_matching, row, start - 1, {
+        end_col = finish,
+        hl_group = highlight,
+        priority = SORTER_HIGHLIGHTS_PRIORITY,
+        strict = false,
+      })
+    end
+  end
 end
 
 function Highlighter:hi_selection(row, caret)
   caret = vim.F.if_nil(caret, "")
   local results_bufnr = assert(self.picker.results_bufnr, "Must have a results bufnr")
 
   a.nvim_buf_clear_namespace(results_bufnr, ns_telescope_selection, 0, -1)
-  a.nvim_buf_add_highlight(results_bufnr, ns_telescope_selection, "TelescopeSelectionCaret", row, 0, #caret)
-
-  a.nvim_buf_set_extmark(
-    results_bufnr,
-    ns_telescope_selection,
-    row,
-    #caret,
-    { end_line = row + 1, hl_eol = conf.hl_result_eol, hl_group = "TelescopeSelection" }
-  )
+
+  -- Skip if there is nothing on the actual line
+  if a.nvim_buf_get_lines(results_bufnr, row, row + 1, false)[1] == "" then
+    return
+  end
+
+  local offset = #caret
+
+  -- Highlight the caret
+  a.nvim_buf_set_extmark(results_bufnr, ns_telescope_selection, row, 0, {
+    virt_text = { { caret, "TelescopeSelectionCaret" } },
+    virt_text_pos = "overlay",
+    end_col = offset,
+    hl_group = "TelescopeSelectionCaret",
+    priority = SELECTION_HIGHLIGHTS_PRIORITY,
+    strict = true,
+  })
+
+  -- Highlight the text after the caret
+  a.nvim_buf_set_extmark(results_bufnr, ns_telescope_selection, row, offset, {
+    end_line = row + 1,
+    hl_eol = conf.hl_result_eol,
+    hl_group = "TelescopeSelection",
+    priority = SELECTION_HIGHLIGHTS_PRIORITY,
+  })
 end
 
 function Highlighter:hi_multiselect(row, is_selected)
   local results_bufnr = assert(self.picker.results_bufnr, "Must have a results bufnr")
 
+  if not a.nvim_buf_is_valid(results_bufnr) then
+    return
+  end
+
+  a.nvim_buf_clear_namespace(results_bufnr, ns_telescope_multiselection, row, row + 1)
+
+  local line = a.nvim_buf_get_lines(results_bufnr, row, row + 1, false)[1]
+  if not line then
```

Comment:
line 108 also does this check. but it compares with `""`, I think it doesnt hurt to check for both nil and "" in 108 and here